math/Line2.cs

```diff
@@ -1,51 +1,50 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
 
 namespace g3
 {
-    public struct Line2d
+    public readonly struct Line2d
     {
-        public Vector2d Origin;
-        public Vector2d Direction;
+        public readonly Vector2d Origin;
+        public readonly Vector2d Direction;
 
-        public Line2d(Vector2d origin, Vector2d direction) {
-            this.Origin = origin;
-            this.Direction = direction;
+        public Line2d(Vector2d origin, Vector2d direction)
+        {
+            Origin = origin;
+            Direction = direction;
         }
 
-        public Line2d(ref Vector2d origin, ref Vector2d direction) {
-            this.Origin = origin;
-            this.Direction = direction;
+        public Line2d(ref Vector2d origin, ref Vector2d direction)
+        {
+            Origin = origin;
+            Direction = direction;
         }
 
-        public static Line2d FromPoints(Vector2d p0, Vector2d p1) {
-            return new Line2d(p0, (p1 - p0).Normalized);
-        }
-        public static Line2d FromPoints(ref Vector2d p0, ref Vector2d p1) {
-            return new Line2d(p0, (p1 - p0).Normalized);
-        }
+        public static Line2d FromPoints(Vector2d p0, Vector2d p1) =>
```

Comment:
in Vector2d

Reply:
There's already a 'ref' overload of this method which prevents me from making the change you suggest.